paasta_tools/iptables.py

```diff
@@ -0,0 +1,113 @@
+"""iptables helper functions.
+
+Unlike the `firewall` module, these functions know nothing about PaaSTA and
+could effectively be a third-party library. They just make working with
+iptables a little bit easier.
+"""
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import re
+import shlex
+import subprocess
+
+
+class ChainDoesNotExist(Exception):
+    pass
+
+
+def all_chains():
+    output = subprocess.check_output(
+        ('iptables', '-t', 'filter', '--list-rules'),
+    )
+    chain_regex = re.compile('-N ([^ ]+)$')
+    chains = set()
+    for line in output.splitlines():
+        m = chain_regex.match(line)
+        if m:
+            chains.add(m.group(1))
+    return chains
+
+
+def ensure_chain(chain, rules=None):
+    """Idempotently ensure a chain exists and has an exact set of rules.
+
+    This function creates or updates an existing chain to match the rules
+    passed in.
+
+    This function will not reorder existing rules, but any new rules are always
+    inserted at the front of the chain.
+    """
+    try:
+        current_rules = list_chain(chain)
+    except ChainDoesNotExist:
+        create_chain(chain)
+        current_rules = set()
+
+    if rules is not None:
+        for rule in rules:
+            if rule not in current_rules:
+                add_rule(chain, rule)
+
+        for rule in current_rules:
+            if rule not in rules:
+                delete_rule(chain, rule)
+
+
+def ensure_rule(chain, rule):
+    rules = list_chain(chain)
+    if rule not in rules:
+        add_rule(chain, rule)
+
+
+def add_rule(chain, rule):
```

Comment:
A bit bikesheddy, but imho since we're putting these rules at the top of the chain rather than appending, call this insert_rule() instead of add_rule()